Provide an iterator for CertificateChain

kim opened the pull request:
Trying to integrate #644 I realise I'm kind of back to square one, because I can't get at the actual certificate :)

Ralith commented on quinn-proto/src/crypto/types.rs:
```diff
@@ -52,6 +52,14 @@ impl CertificateChain {
+            .iter()
+            .cloned()
+            .map(|cert| Certificate { inner: cert })
+    }
```
I'd prefer for this to either be turned into `into_iter` (consuming `self`) or to finagle things until it can become an `impl Iterator<Item = &'a Certificate>`.

djc replied:
Why is that? Implementing `Iterator` does not seem to make sense here.

Ralith replied:
The trait associated with `into_iter` is `IntoIterator`.

djc replied:
Yeah, wasn't sure about what you meant by "finagle things until it can become an `impl Iterator`. So you mean `impl IntoIterator for &CertificateChain`? That sounds like a plan.

kim replied:
I'm not sure I understand how I could turn an `impl Iterator<Item = &'a rustls::Certificate>` into an `impl Iterator<Item = &'a Certificate>`

Ralith replied:
I meant `impl IntoIterator for CertificateChain`, or `fn iter(&self) -> impl Iterator<Item = &Certificate>`, as two reasonable alternatives.

Ralith replied:
The latter option would involve storing a `Vec<Certificate>` or similar internally and converting for rustls's convenience as necessary. But I think the former is a good choice if it serves your needs.

kim replied:
Ok thanks.

I was hesitant to make swiping changes, as `CertificateChain::certs` is `pub(crate)`. If that’s appreciated, I’ll try to be a bit more audacious next time :)

djc commented:
Maybe we should also try to upstream some of the more friendly interfaces from our wrappers into rustls? The current solution feels a bit kludgey.

Ralith commented:
AFAIK the main benefit of these interfaces as they stand is allowing people to avoid exposure to rustls's unstable API, so I'm not sure that would b a win.

djc commented:
> AFAIK the main benefit of these interfaces as they stand is allowing people to avoid exposure to rustls's unstable API, so I'm not sure that would be a win.

Well, my point is whether we can make a similar API stable in rustls. I filed https://github.com/ctz/rustls/issues/348, feel free to comment there if you have any ideas around this.

djc commented:
@kim, both of your versions implemented iterators that yield owned `Certificate`s. Is there a reason you're staying away from just yielding `&Certificate`s instead? I think that would be preferable here (for one thing, so the `AuthenticationData` can stay intact).

kim commented:
@djc Because `CertificateChain` stores `rustls::Certificate`s internally, but that field is also `pub(crate)`, I was afraid of having to introduce a lot of conversion boilerplate if I changed that to store quinn's `Certificate` instead (which is the only way I can see implementing `impl Iterator<Item = &Certificate>` can work).

It turns out that only one callsite is affected (or even exists?), so it's not too bad.

I'm not really familiar with the `impl<'a> Move for &'a Ref` idiom, and also find it confusing when a simple `fn iter(&self) -> impl Iterator<Item = &Ref>` would do, so I implemented the latter in 015c498bee74438da0d6d906b52842f1211f290f

djc commented:
I'm sorry, I meant to distinguish between iterating over references vs iterating over owned types. I think the current changes probably contain too much stuff. Can we just have an `impl IntoIterator<Item = &rustls::Certificate> for &CertificateChain`? Would that solve your use case?

kim commented:
Sure. You prefer `impl IntoIterator` over just `fn iter()`?

Ralith commented:
> Well, my point is whether we can make a similar API stable in rustls.

That would still lead to confusing breakage on each rustls update due to crate version mismatches, unless we reexport rustls per #300 and encourage people to use it that way.

> Sure. You prefer impl IntoIterator over just fn iter()?

Idiomatically you'd have both.

kim commented:
I feel like we're talking past each other. To summarise:

* #644 generalises #638 to provide access to data negotiated during the cryptographic handshake, as

kim commented:
I feel like we're talking past each other. To summarise:

* #567 requested to provide access to data negotiated during the cryptographic handshake
* #594 implements this, but only on the `quinn-proto` layer
* #638 proposes to expose access to peer certificates on the `quinn` layer as well
* #644 generalises this, the intention is to not only abstract over the cryptographic protocol, but also hide the implementation detail of `rustls`
* Unfortunately, this actually breaks the original motivation, because the `CertificateChain` newtype doesn't provide access to the underlying certificate data. Not being too familiar with the codebase, I missed this during review, sorry about that. This PR is meant to rectify the situation.
* My assumption about the newtypes in `quinn-proto::crypto::types` was that they are there to _hide_ `rustls`. This seems to be wrong, their purpose is to add convenience methods. I do agree with @burdges that this might be better achieved using "Ext" traits.
* Alright, so it's ok to provide an iterator over `rustls::Certificate`s, and not having to wrap it in the newtype again.
* That's cool, but then I don't get how `impl IntoIterator for &CertificateChain` is useful or idiomatic -- it's the same thing as `self.certs.iter()`, only gives the compiler more work. My understanding is that `Into`/`into` is a convention to indicate that the receiver will be consumed. But maybe I'm missing something.

Alright. So if I provide a `impl IntoIterator for CertificateChain`, would you be willing to accept this patch?

kim commented:
> So if I provide a `impl IntoIterator for CertificateChain`, would you be willing to accept this patch?

Sorry, to clarify: `fn iter(&self) -> impl Iterator<Item = rustls::Certificate>` and `impl IntoIterator for CertificateChain` where `type Item = rustls::Certificate`.

Ralith commented:
I think me and @djc have somewhat different perspectives here, and that's produced some confusion. I apologize for that.

> My assumption about the newtypes in quinn-proto::crypto::types was that they are there to hide rustls.

This was my intention in adding those newtypes to begin with, though the specific goal was to *allow* callers to be ignorant of rustls, not necessarily to force them to--so exposing the details is acceptable so long as it's not required for  simple cases.

> I don't get how impl IntoIterator for &CertificateChain is useful or idiomatic -- it's the same thing as self.certs.iter(), only gives the compiler more work.

`IntoIterator` on references is the a common idiom because it makes `for x in &xs` work, which is a bit more pleasant than `for x in xs.iter()`. You'll see this pattern in every std collection, e.g. for slices and `Vec`.

> So if I provide a impl IntoIterator for CertificateChain, would you be willing to accept this patch?

I'd be happy to merge either/both of those, though I'd consider an approach which doesn't expose rustls to be better still. An easier solution than wrapping/unwrapping the newtypes might be to fall back on `Vec<u8>` for owning iterators and `&[u8]` for borrowing ones.

kim commented:
> IntoIterator on references is the a common idiom because it makes for x in &xs work

Ok, thanks. Please forgive my ignorance, I'm coming from a programming language whose main compiler can auto-derive _every_ typeclass through a newtype 🤷‍♂

Ralith commented:
No worries, always happy to help someone get up to speed!

djc commented:
Yeah, sorry for the confusion and mixed messages here, and thanks for sticking with it!
